---
title: "Dismantling the bulk: examining neuronal heterogeneity using single-cell techniques"
author: Sara Linker, Apua Paquola, Roger Lasken, and Keegan Korthauer
date: 8/30/2016
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(out.width='750px', out.height='750px', dpi=300,
                      fig.height=7, fig.width=7)
knitr::opts_knit$set(root.dir="~/FestivalWorkshopSC/BrainAtlas")
```

### Welcome to the Festival of Genomics workshop on single-cell analyses

This is an R Markdown document that contains instructions and code for the examples used in todays workshop.  The first few steps will check that you have all the packages and data files necessary to carry out all of the analyses.

# Hour 1: Getting Started
Sara's section
...

### Check that the Brain Atlas data files are present

The following code chunk assumes that Brain Atlas files have been downloaded and placed in the "BrainAtlas" subdirectory of a folder in your home directory entitled "FestivalWorkshopSC".  If you have downloaded these files to another location, either create a new folder and move the files there, or substitute the file path to where they are currently located for "~/FestivalWorkshopSC/BrainAtlas".  If you are using the RStudio Server instance provided by the workshop, change the first line that follows to `setwd("/home/FestivalWorkshopSC/BrainAtlas")`

```{r Check for data files, eval=TRUE, echo=TRUE}
setwd("~/FestivalWorkshopSC/BrainAtlas")
file.exists("cell_metadata.csv")
file.exists("genes_counts.csv")
file.exists("README.txt")
```

If any of the preceding lines return `FALSE`, double check that you have set the correct working directory and that all download files have been placed in that folder.  If they are missing, you can download the files [here](http://casestudies.brain-map.org/celltax/download_archive).  Note that there will be more files than listed here (including alternate gene count quantifications, and metadata about data-driven cluster memberships as discussed in the paper ["Adult mouse cortical cell taxonomy revealed by single cell transcriptomics"](http://www.nature.com/neuro/journal/v19/n2/full/nn.4216.html)), but these are the ones we will make use of.

### Read the Brain Atlas data files into R

The `read.csv` function in R is useful for reading in .csv (comma-separated value) files.  First, we'll read in the main data file `genes_counts.csv` to a data frame using this function and check its contents.  The extra arguments to this function help to format our object so that we have row and column names, and we use character variables instead of converting to factors.  For more details on these arguments, you can type `help(read.csv)`.  In the resulting `coutns` object, we have genes in rows (24057) and cells in columns (1679).

```{r Read in Counts, eval=TRUE, echo=TRUE}
counts <- read.csv("genes_counts.csv", stringsAsFactors = FALSE, header=TRUE, row.names = 1)
str(counts[,1:20]) # restrict to the first 20 columns (cells) 
```

The `cell_metadata.csv` file contains 1679 rows (one for each cell) and columns containing information such as collection date, sequencing type, total reads, mapping percentage, dissection layer, and major/minor derived cell subtypes. 

```{r Read in Cell Metadata, eval=TRUE, echo=TRUE}
cells <- read.csv("cell_metadata.csv", stringsAsFactors = FALSE, header = TRUE)
str(cells)
```

More detailed information about the files downloaded from the Allen Brain Atlas can be found in the `README.txt` file provided.  Here is a peek at the contents of that file.

```{bash Peek at README.txt file, eval=TRUE, echo=TRUE}
# This is a bash command, to be executed at the command line (not within R);
# Alternatively, simply open the README.txt in your favorite text editor to view its contents
head README.txt
```

### Check that the desired R packages have been installed

Once a list of desired R packages is finalized, can check that they are installed with 

```{r Check for desired packages, eval=TRUE, echo=TRUE, results="hide", message=FALSE, warning=FALSE}
require(scde)     #bioconductor
require(monocle)  #bioconductor
require(sincell)  #bioconductor
require(scDD)     #github
require(ggplot2)  #cran
require(devtools) #cran
```

If any of these commands return a message that includes "there is no package called...", then the package is missing and needs to be installed.  Note that packages may be stored in one of several package repositories.  The most popular are Bioconductor, github, and CRAN.  For Bioconductor packages, for example ```edgeR```, this can be done with the following code:

```{r install bioconductor package, echo=TRUE, eval=FALSE, results="hide", message=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("monocle")
```

For CRAN packages, for example ```devtools```, installation can be done with the following code:

```{r install cran packages, echo=TRUE, eval=FALSE}
install.packages(devtools)
``` 

For Github packages, for example ```scDD```, installation can be done with the following code:

```{r install github packages, echo=TRUE, eval = FALSE}
install.packages("devtools")
devtools::install_github("kdkorthauer/scDD")
```

### Visualize major axes of variation in a PCA plot

```{r PCA, eval = TRUE, echo = TRUE}
# extract top 500 variable genes
gene.var <- apply(counts, 1, function(x) var(log(x[x>0])))
counts.top500 <- counts[which(rank(-gene.var)<=500),]

counts.pca <- prcomp(log(counts.top500+1),
                   center = TRUE,
                   scale. = TRUE) 
summary(counts.pca)$importance[,1:5]
plot(counts.pca, type="l", main="Top 10 PCs")

color_class <- rainbow(length(unique(cells$major_class)))
plot(counts.pca$rotation[,1], counts.pca$rotation[,2], 
      xlab="PC 1", ylab="PC 2", col=color_class[as.numeric(factor(cells$major_class))], pch=20,
      main="PCA plot of cells colored by derived major class")

color_class <- rainbow(length(unique(cells$layer_dissectoin)))
plot(counts.pca$rotation[,1], counts.pca$rotation[,2], 
      xlab="PC 1", ylab="PC 2", col=color_class[as.numeric(factor(cells$layer_dissectoin))], pch=20,
      main="PCA plot of cells colored by Dissection Layer")
```

# Hour 2: Normalization and Quality Control of scRNA-seq
Apua's Section 
...

### Normalization

...

### Quality Control (QC measures)

```{r Detection Rate, eval = TRUE, echo = TRUE}
detectionRate <- apply(counts, 2, function(x) sum(x > 0) / length(x))
hist(detectionRate)
```

...


# Hour 3: Analysis Modules

Keegan's Section
...

### Identify the highly variable genes

```{r Highly Variable Genes, eval = TRUE, echo = TRUE}
library(sincell)
```
...

### Identify differentially expressed genes

```{r DE, eval = TRUE, echo = TRUE}
library(scde)
library(scDD)
```
...

### Order cells by "Psuedotime" (temporal-spatial variation)

```{r Pseudotime, eval = TRUE, echo = TRUE}
library(monocle)
```
...

### Identify genes with oscillating expression patterns

```{r Oscillating genes, eval = TRUE, echo = TRUE}
library(Oscope)
```

TODO: Fill in the above analysis modules ...

```{r Extract Code Snippets, eval = FALSE, echo = FALSE}
# This snippet just generates a .R file that only contains the code snippets within this document
# Not executed; need to run separately to update the .R file after this file is modified
library(knitr)
knitr:::purl("~/Desktop/scRNAseq/FestivalWorkshop2016/FestivalWorkshopVignettes/SingleCellAnalyses.Rmd",  
     output="~/Desktop/scRNAseq/FestivalWorkshop2016/FestivalWorkshopVignettes/SingleCellAnalyses.R")
```
