Dismantling the bulk: examining neuronal heterogeneity using single-cell techniques
================
Sara Linker, Apua Paquola, Roger Lasken, and Keegan Korthauer
7/25/2016

Welcome to the Festival of Genomics workshop on single-cell analyses
--------------------------------------------------------------------

This is an R Markdown document that contains instructions and code for the examples used in todays workshop. The first few steps will check that you have all the packages and data files necessary to carry out all of the analyses.

Getting Started
---------------

### Check that the Brain Atlas data files are present

The following code chunk assumes that Brain Atlas files have been downloaded and placed in a folder in your home directory entitled "FestivalWorkshopSC". If you have downloaded these files to another location, either create a new folder and move the three .csv data files and one documentation text file there, or substitute the file path to where they are currently located for "~/FestivalWorkshopSC".

``` r
setwd("~/FestivalWorkshopSC/")
file.exists("columns-cells.csv")
```

    ## [1] TRUE

``` r
file.exists("fpkm_table.csv")
```

    ## [1] TRUE

``` r
file.exists("rows-genes.csv")
```

    ## [1] TRUE

``` r
file.exists("README.txt")
```

    ## [1] TRUE

If any of the preceding lines return `FALSE`, double check that you have set the correct working directory and that all four download files have been placed in that folder. If they are missing, you can download the files [here](http://celltypes.brain-map.org/api/v2/well_known_file_download/502999251).

### Read the Brain Atlas data files into R

The `read.csv` function in R is useful for reading in .csv (comma-separated value) files. First, we'll read in the main data file `fpkm_table.csv` to a data frame using this function and check its contents. The extra arguments to this function help to format our object so that we have row and column names, and we use character variables instead of converting to factors. For more details on these arguments, you can type `help(read.csv)`. In the resulting `fpkm` object, we have genes in rows (45,771) and cells in columns (1772).

``` r
fpkm <- read.csv("fpkm_table.csv", stringsAsFactors = FALSE, header=TRUE, row.names = 1)
str(fpkm)
```

    ## 'data.frame':    45771 obs. of  1772 variables:
    ##  $ X497122137: num  6.87 30.38 0.03 0 1.82 ...
    ##  $ X497122141: num  17.61 0 0 0 4.91 ...
    ##  $ X497122142: num  0 0 0 0 2.06 0 0 0 0 0 ...
    ##  $ X497122143: num  0.35 0 1.22 0 3.46 0 0.23 0 0 0 ...
    ##  $ X497122144: num  1.32 0 0.35 0 2.57 0 0.13 0 0 0 ...
    ##  $ X497122145: num  0.69 0 0.58 0 2.75 0 0 0 0 0 ...
    ##  $ X497122146: num  0.44 0 0.38 0 2.47 0 0 0 0 0 ...
    ##  $ X497122147: num  0.31 0 2.99 0 7.36 0 0 0 0 0 ...
    ##  $ X497122148: num  8.31 0 2.59 0 6.6 0 0.44 0 0 0 ...
    ##  $ X497122149: num  5.56 0 0 0 0.65 0 0 0 0 0 ...
    ##  $ X497122150: num  0.31 0 8 0 4.08 0 0 0 0 0 ...
    ##  $ X497122151: num  13.79 0 7.75 0 0.44 ...
    ##  $ X497122152: num  13.9 12.8 14.5 12.2 1.3 ...
    ##  $ X497122153: num  24.36 0 6.58 0 12.04 ...
    ##  $ X497122154: num  13.44 0 9.48 0 0 ...
    ##  $ X497122155: num  9.54 0 1.62 0 1.54 0 0 0 0 0 ...
    ##  $ X497122156: num  7.1 0 0.45 0 3.47 0 0 0 0 0 ...
    ##  $ X497122157: num  1.65 0 0 0 1.3 0 0 0 0 0 ...
    ##  $ X497122158: num  7.89 0 5.45 0 8.47 0 0 0 0 0 ...
    ##  $ X497122159: num  9.85 0 12.89 0 0.74 ...
    ##  $ X497122160: num  3.09 0 5.63 0 14.12 ...
    ##  $ X497122161: num  7.54 0 17.22 0 3.52 ...
    ##  $ X497122162: num  2.85 0 0.19 0 0.04 0 0 0 0 0 ...
    ##  $ X497122163: num  2.22 0 0.23 0 1.12 0 0 0 0 0 ...
    ##  $ X497122164: num  0 0 0 0 6.55 0 0 0 0 0 ...
    ##  $ X497122165: num  4.26 0 0 0 0 0 0.05 0 0 0 ...
    ##  $ X497122166: num  0.05 0 0 0 0.48 0 0 0 0 0 ...
    ##  $ X497122167: num  0.71 0 0 0 6.04 0 0 0 0 0 ...
    ##  $ X497122168: num  0 0 0 0 0.2 0 0 0 0 0 ...
    ##  $ X497122169: num  4.2 0 0 0 0.21 0 0 0 0 0 ...
    ##  $ X497122170: num  12.89 5.21 0.43 0 3.47 ...
    ##  $ X497122171: num  9.27 0 0 0 1.93 0 0.74 0 0 0 ...
    ##  $ X497122172: num  9.97 0 1.66 0 0.14 0 0 0 0 0 ...
    ##  $ X497122174: num  0 0 6.43 0 0 0 0 0 0 0 ...
    ##  $ X497122175: num  11.4 0 0 0 0 ...
    ##  $ X497122176: num  7.71 0 0 0 1 0 0 0 0 0 ...
    ##  $ X497122177: num  1.56 0 0.06 0 0.04 1.72 0 0 0 0 ...
    ##  $ X497122178: num  0 0 17.25 0 9.71 ...
    ##  $ X497122179: num  14.07 0 5.71 0 3.39 ...
    ##  $ X497122180: num  2.43 0 0.81 0 0.37 0 0 0 0 0 ...
    ##  $ X497122181: num  0.19 0 3.62 0 0.38 0 0 0 0 0 ...
    ##  $ X497122182: num  0 0 0 0 0 0 0 0 0 0 ...
    ##  $ X497122183: num  2.6 107.24 0.14 0 7.65 ...
    ##  $ X497122184: num  1.78 0 0.41 0 8.46 0 0 0 0 0 ...
    ##  $ X497122185: num  3.1 0 0 0 0.84 0 0 0 0 0 ...
    ##  $ X497122188: num  8.82 0 0 0 2.78 0 0 0 0 0 ...
    ##  $ X497122189: num  0.34 0 0 0 0 0 0 0 0 0 ...
    ##  $ X497122190: num  0 0 0 0 1.57 0 0 0 0 0 ...
    ##  $ X497122191: num  33.7 0 3.68 0 15.98 ...
    ##  $ X497122192: num  0.65 0 4.52 0 1.64 0 0 0 0 0 ...
    ##  $ X497122193: num  14.6 0 0 0 3.07 0 0 0 0 0 ...
    ##  $ X497122194: num  23.2 0 1.2 0 2.5 ...
    ##  $ X497122195: num  6.95 0 0.73 0 0.06 0 0 0 0 0 ...
    ##  $ X497122196: num  12.03 0 0 0 2.65 ...
    ##  $ X497122197: num  5.88 0 3.37 0 4.69 0 0 0 0 0 ...
    ##  $ X497122198: num  4.53 0 5.67 0 1.31 0 0 0 0 0 ...
    ##  $ X497122199: num  2.11 0 2.29 0 5.15 0 0 0 0 0 ...
    ##  $ X497122200: num  0 0 0.77 0 0.46 0 0 0 0 0 ...
    ##  $ X497122201: num  4.28 0 0 0 0.81 0 0 0 0 0 ...
    ##  $ X497122202: num  0 0 7.45 0 12.1 0 0 0 0 0 ...
    ##  $ X497122203: num  54.27 0 6.97 0 14.51 ...
    ##  $ X497122204: num  4.8 0 0.38 0 0.04 0 0 0 0 0 ...
    ##  $ X497122205: num  16.97 0 0.1 0 1.87 ...
    ##  $ X497122206: num  9.35 0 0 0 0.45 0 0 0 0 0 ...
    ##  $ X497122207: num  8.85 0 6.61 0 0 0 0.19 0 0 0 ...
    ##  $ X497122208: num  3.52 0 0 0 0.21 0 0 0 0 0 ...
    ##  $ X497122209: num  1.45 0 0.11 0 0.83 0 0.43 0 0 0 ...
    ##  $ X497122210: num  13.14 0 0 0 0.96 ...
    ##  $ X497122211: num  5.24 0 0.56 0 0.09 0 0 0 0 0 ...
    ##  $ X497122212: num  18.75 0 15.51 0 4.14 ...
    ##  $ X497122213: num  11.13 0 4.29 0 9.76 ...
    ##  $ X497122214: num  1.12 0 6.6 0 0.09 0 0 0 0 0 ...
    ##  $ X497122215: num  26.17 0 15.45 0 5.56 ...
    ##  $ X497122216: num  0 0 0 0 1.19 0 0 0 0 0 ...
    ##  $ X497122217: num  17.29 0 2.14 0 6.81 ...
    ##  $ X497122218: num  13.8 0 2.95 0 11.82 ...
    ##  $ X497122219: num  13.42 0 9.14 0 24.29 ...
    ##  $ X497122220: num  11.22 0 4.43 0 6.07 ...
    ##  $ X497122221: num  7.16 0 4.5 0 1.65 0 0 0 0 0 ...
    ##  $ X497122222: num  24.29 0 2.66 0 10.09 ...
    ##  $ X497122223: num  14.39 0 3.01 0 15.57 ...
    ##  $ X497122224: num  14.72 0 0.23 0 15.74 ...
    ##  $ X497122225: num  30.79 0 0.63 0 8.63 ...
    ##  $ X497122226: num  18.5 0 22.1 0 21.4 ...
    ##  $ X497122227: num  17.76 0 2.56 0 6.6 ...
    ##  $ X497122236: num  27.95 0 8.25 0 4.3 ...
    ##  $ X497122237: num  39.68 0 4.36 0 7.96 ...
    ##  $ X497122238: num  25.46 0 12.35 0 8.83 ...
    ##  $ X497122239: num  12.17 0 2.6 0 6.42 ...
    ##  $ X497122240: num  26.93 0 19.74 0 8.21 ...
    ##  $ X497122241: num  12.74 0 0.23 5.63 1.83 ...
    ##  $ X497122242: num  17.3 0 6.84 0 3.26 0 0.12 0 0 0 ...
    ##  $ X497122243: num  31.51 0 10.36 0 7.89 ...
    ##  $ X497122244: num  15.87 0 6.08 0 13.64 ...
    ##  $ X497122245: num  5.08 0 8.2 0 2.37 0 0 0 0 0 ...
    ##  $ X497122246: num  12.02 0 1.66 0 15.55 ...
    ##  $ X497122247: num  4.98 0 10.62 0 5.77 ...
    ##  $ X497122248: num  20.6 0 11.2 0 10.2 ...
    ##  $ X497122249: num  25.69 0 11.17 0 7.92 ...
    ##   [list output truncated]

Next, we'll read in the metadata files, which contain additional information about the rows and columns of the `fpkm` object. The `rows-genes.csv` file contains 45,771 rows (one for each gene) and 5 columns containing information such as the chromosomal location, gene symbol, and gene description.

``` r
genes <- read.csv("rows-genes.csv", stringsAsFactors = FALSE, header = TRUE)
str(genes)
```

    ## 'data.frame':    45771 obs. of  5 variables:
    ##  $ gene_id       : int  500699669 500699670 500699671 500699672 500699673 500699674 500699675 500699676 500699677 500699678 ...
    ##  $ chromosome    : chr  "1" "1" "1" "1" ...
    ##  $ gene_entrez_id: int  497097 100418032 105243853 664792 105242387 105242467 619785 19888 20671 664830 ...
    ##  $ gene_symbol   : chr  "Xkr4" "Gm18956" "LOC105243853" "Gm7341" ...
    ##  $ gene_name     : chr  "X Kell blood group precursor related family member 4" "predicted gene, 18956" "uncharacterized LOC105243853" "predicted gene 7341" ...

The \`columns-cells.csv' file contains 1772 rows (one for each cell) and columns containing information such as age of the donor mouse, targeted brain region, brain slice ID, percentage of reads that mapped, and putative cell subtype.

``` r
cells <- read.csv("columns-cells.csv", stringsAsFactors = FALSE, header = TRUE)
str(cells)
```

    ## 'data.frame':    1772 obs. of  22 variables:
    ##  $ rnaseq_profile_id                                  : int  497122137 497122141 497122142 497122143 497122144 497122145 497122146 497122147 497122148 497122149 ...
    ##  $ donor_id                                           : int  485247214 485247214 485247214 485247214 485247214 485247214 485247214 485247214 485247214 485247214 ...
    ##  $ donor_age                                          : chr  "P54" "P54" "P54" "P54" ...
    ##  $ genotype_driver                                    : chr  "Slc32a1-IRES-Cre" "Slc32a1-IRES-Cre" "Slc32a1-IRES-Cre" "Slc32a1-IRES-Cre" ...
    ##  $ genotype_reporter                                  : chr  "Ai14" "Ai14" "Ai14" "Ai14" ...
    ##  $ cell_reporter                                      : chr  "cre reporter positive" "cre reporter positive" "cre reporter positive" "cre reporter positive" ...
    ##  $ cell_prep_sample_id                                : int  486509295 486509295 486509295 486509295 486509295 486509295 486509295 486509295 486509624 486509624 ...
    ##  $ sampling_region                                    : chr  "Mouse LGN - Core" "Mouse LGN - Core" "Mouse LGN - Core" "Mouse LGN - Core" ...
    ##  $ slice_index                                        : int  11 11 11 11 11 11 11 11 13 13 ...
    ##  $ slice_id                                           : int  486749503 486749503 486749503 486749503 486749503 486749503 486749503 486749503 486749564 486749564 ...
    ##  $ slice_name                                         : chr  "Slc32a1-IRES-Cre;Ai14-204789.11.01" "Slc32a1-IRES-Cre;Ai14-204789.11.01" "Slc32a1-IRES-Cre;Ai14-204789.11.01" "Slc32a1-IRES-Cre;Ai14-204789.11.01" ...
    ##  $ cell_id                                            : int  501389972 501389973 501389974 501389975 501389976 501389977 501389978 501389979 501389980 501389981 ...
    ##  $ rnaseq_profile_total_reads                         : int  3380643 3252488 3309330 3576681 4954087 3920690 3859904 2221764 617242 718808 ...
    ##  $ rnaseq_profile_percent_reads_aligned_to_mrna       : num  48.7 51.1 48.8 66.7 52.2 45.3 60.7 63.5 63.4 61.8 ...
    ##  $ rnaseq_profile_percent_reads_aligned_to_ncrna      : num  2.05 3.55 3.63 2.48 3.88 4.01 2.84 3.01 3.05 2.9 ...
    ##  $ rnaseq_profile_percent_reads_aligned_to_genome_only: num  9.35 16.4 17.7 10.4 16.1 20.4 13.6 12.1 11.5 8.5 ...
    ##  $ subclass                                           : chr  "Styk1_Xylb" "Styk1_Cdh13_Car10" "Tac1_Gabra4" "Styk1_Xylb" ...
    ##  $ subclass_order                                     : int  9 7 6 9 7 6 9 9 18 18 ...
    ##  $ subclass_color                                     : chr  "#FF3030" "#FF6A6A" "#CE1256" "#FF3030" ...
    ##  $ broad_class                                        : chr  "Gad2_Chrna6" "Gad2_Chrna6" "Gad2_Chrna6" "Gad2_Chrna6" ...
    ##  $ broad_class_order                                  : int  3 3 3 3 3 3 3 3 4 4 ...
    ##  $ broad_class_color                                  : chr  "#FF0000" "#FF0000" "#FF0000" "#FF0000" ...

### Check that the desired R packages have been installed

Once a list of desired R packages is finalized, can check that they are installed with

``` r
require(edgeR)
require(scde)
require(monocle)
```

If any of these commands return a message that includes "there is no package called...", then the package is missing and needs to be installed. For Bioconductor packages, for example `edgeR`, this can be done with the following code:

``` r
source("http://bioconductor.org/biocLite.R")
biocLite("monocle")
```

TODO: add analyses modules ...
